kafkaesk/app.py

```diff
@@ -236,22 +254,21 @@ async def __run(self) -> None:
                         partition=record.partition,
                         group_id=self._subscription.group,
                     ).inc()
-                except UnhandledMessage:
-                    # how should we handle this? Right now, fail hard
-                    logger.warning(f"Could not process msg: {record}", exc_info=True)
-                    # We increase the error counter
+                except Exception as err:
                     CONSUMED_MESSAGES.labels(
                         stream_id=record.topic,
                         partition=record.partition,
-                        error="UnhandledMessage",
+                        error=err.__class__.__name__,
                         group_id=self._subscription.group,
                     ).inc()
-
+                    await retry_policy(record=record, error=err)
                 finally:
                     context.span.finish()
                     context.close()
                     await self.emit("message", record=record)
         finally:
+            # Shutdown the retry policy
+            await retry_policy.finalize()
             try:
                 await self._consumer.commit()
```

Comment:
i still trying to figure out how we should couple retry logic with the fact that we should or should not commit, maybe something like:

```
# or either commiting in the retry policy itself at finalizing time
if await retry_policy.finalize_and_commit():
   await self._consumer.commit()
```

@vangheem @tysondenio do you have any thoughts?

Reply:
If I'm understanding everything correctly this `finally` block is run outside of the main consumption loop.  This is the code that is run when the consumer is terminated.

It looks like we are using `auto_commit` (enabled by default) to manage commits for each individual message in the main queue.  I think this is the desired behavior and holds true with some assumptions I made with the design.

Currently, my thought is that we would not be re-processing messages by reading the same message instance again from the main stream (by means of not committing the offset and allowing the consumer to pull the same message again).  In order for a message to be re-processed a copy of the message needs to be explicitly re-published to a stream and re-consumed.  Since our main streams are shared across components we will need to maintain separate streams for managing these messages.

# Example:
1. `SubscriptionConsumer` reads a message from `main_stream`.
2. `SubscriptionConsumer` parses message and calls `Consumer.func()` _(user supplied callback)_.
3. `Consumer.func()` raises an `Exception`
4. `SubscriptionConsumer` catches `Exception` and passes message and `Exception` to `RetryPolicy.handle()`.
5. `RetryPolicy.handle()` finds a match for `Exception` and calls `RetryHandler()`.
6. `RetryHandler()` publishes a new `RetryMessage` consisting of the message and any `RetryHistory` to `retry_stream`.

Later, possibly on a different instance:

1. `RetryHandler.RetryConsumer` _(no examples of this ATM)_ reads a message from `retry_stream`.
2. `RetryHandler.RetryConsumer` parses the message and calls `RetryPolicy.subscription.func()`.
3. If `RetryPolicy.subscription.func()` raises an exception again, return to step 4 above.  This loop will continue until the message is successful or the `RetryPolicy` decides not to forward the message to `retry_stream`.

I will confess that I haven't looked very deeply into attempting retries with offset manipulation but I don't believe that approach would solve all of our problems:

1. My initial impression is that the retry attempts would block / impact our ability to continue processing live data.
2. Implementing features like backoff timers would likely become more difficult.
3. Retry will always happen on the same consumer.  Instance-based issues will continually cause the message to fail even if it would process fine on another instance.

Please let me know if I'm missing something here

Reply:
All these handled exceptions would commit.

IMO, hard failures are the only types of messages that should not commit. These are exceptions that should bubble and cause hard failures.

Otherwise, all errors should forward messages to other topics and/or explicitly drop the message to ignore the error.

Reply:
ok, my understanding is that as the code is now we may loss data when the consumer ends in a unhandled manner, those messages right now are commited anyway, either by the auto_commit logic or because we are commiting when finalizing the consumer, i wrote this little test to have a better understanding:

```
async def test_subscription_failure(app):
    probe = Mock()

    @app.schema("Foo", version=1)
    class Foo(pydantic.BaseModel):
        bar: str

    @app.subscribe("foo.bar", group="test_group")
    async def noop(data: Foo):
        probe("error", data)
        raise Exception("Unhandled Exception")

    async with app:
        fut = asyncio.create_task(app.consume_for(1, seconds=5))
        await asyncio.sleep(0.2)

        await app.publish("foo.bar", Foo(bar=1))
        await app.flush()
        await fut

    # remove wrong consumer
    app._subscriptions = []


    @app.subscribe("foo.bar", group="test_group")
    async def noop_ok(data: Foo):
        probe("ok", data)

    async with app:
        fut = asyncio.create_task(app.consume_for(2, seconds=5))
        await asyncio.sleep(0.5)

        await app.publish("foo.bar", Foo(bar=2))
        await app.flush()

        await fut
```

current output:
```
(Pdb++) probe.mock_calls
[call('error', Foo(bar='1')), call('ok', Foo(bar='2'))]
# first message is lost since has been commited
```


disabling auto_commit, and commiting on close:
```
(Pdb++) probe.mock_calls
[call('error', Foo(bar='1')),
 call('ok', Foo(bar='2')),
 call('ok', Foo(bar='1'))]
# first message is succesfully reprocessed by the updated consumer
```

I think the current behavior is not a sane default in my opinion we should not loss data by default, once default retry policies are in place i think the default should be to not commit on unhandled failure.

Reply:
@dmanchon has identified a valid issue here with how aiokafka's `auto-commit` handles committing offsets when an exception is raised in the `messages` iterator.

Addressing this issue will be handled by implementing custom commit logic in another PR.